paris_rollins.py

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/python2.6
+
+# Run up to a configurable limit of simultaneous paris-traceroutes, back
+# towards recent M-Lab client IP addresses told to us via SideStream.
+
+# TODO(joshb): this script is written to minimize external dependencies.
+# Later versions of subprocess include built in timeout handling, but we
+# can't use them because the M-Lab platform doesn't have them. This should
+# be revisited if the M-Lab platform is upgraded.
+
+# TODO(joshb): this version just provides a function to run paris-traceroute.
+# Next step is to use multiprocessing to run multiple traceroutes and poll
+# the Web100 agent 
+
+import os
+import multiprocessing
+import subprocess
+import sys
+import time
+
+# What binary to use for paris-traceroute
+PARIS_TRACEROUTE = '/usr/local/bin/paris-traceroute'
+# paris-traceroute is run at this nice level, to minimize impact on the host.
+WORKER_NICE = 19
+# paris-traceroute should take no longer than this to complete (timed out,
+# partial results will be discarded).
+WORKER_TIMEOUT = 20 
+# Maximum number of paris-traceoutes to run simultaneously (requests to run
+# more will be discarded).
+MAX_WORKERS = 10
+
+
+def log_worker(message):
+  print time.strftime('%Y%m%d %T %%s', time.gmtime(time.time())) % message
+
+
+def make_log_file_name(log_time, log_file_root, remote_ip, remote_port,
+                       local_ip, local_port):
+  log_file_relative = (time.strftime('%Y/%m/%d/%Y%m%dT%TZ%%s.paris',
+                                     time.gmtime(log_time)) %
+                       ('-'.join((remote_ip, str(remote_port),
+                                  local_ip, str(local_port)))))
+  log_file = os.path.join(log_file_root, log_file_relative)
+  return log_file
+
+
+# Try to run paris-traceroute and log output to a file. We assume any
+# errors are transient (Eg, temporarily out of disk space), so do not
+# crash if the run fails.
+def run_worker(log_file_root, log_time, remote_ip, remote_port,
+               local_ip, local_port):
+  os.nice(WORKER_NICE)
+  command = (
+    '/usr/bin/timeout',
+    str(WORKER_TIMEOUT) + 's',
+    '/usr/local/bin/paris-traceroute',
```

Comment:
Specifying an absolute path name is critical to avoiding security problems
like unnecessary use of a shell and running another process unexpectedly as
root (which this script runs as). It should also be very clear, when the
script runs, exactly what it runs, rather than running whatever it found
first based on the parent script's environment.


On Wed, Aug 19, 2015 at 4:46 PM, Peter Boothe <notifications@github.com>
wrote:

> In paris_rollins.py
> <https://github.com/npad/sidestream/pull/15#discussion_r37430549>:
>
> > +                       ('-'.join((remote_ip, str(remote_port),
> > +                                  local_ip, str(local_port)))))
> > +  log_file = os.path.join(log_file_root, log_file_relative)
> > +  return log_file
> > +
> > +
> > +# Try to run paris-traceroute and log output to a file. We assume any
> > +# errors are transient (Eg, temporarily out of disk space), so do not
> > +# crash if the run fails.
> > +def run_worker(log_file_root, log_time, remote_ip, remote_port,
> > +               local_ip, local_port):
> > +  os.nice(WORKER_NICE)
> > +  command = (
> > +    '/usr/bin/timeout',
> > +    str(WORKER_TIMEOUT) + 's',
> > +    '/usr/local/bin/paris-traceroute',
>
> Absolute pathnames and specifying a specific path for an executable makes
> my eyebrow twitch (Why /usr/local/bin and not /local/bin or /usr/bin or
> /bin? The distinction between /sbin /bin /usr/bin and the like was forced
> on us by now irrelevant max-files-per-directory limitations in early
> versions of Unix and have existed since then as distinctions without a real
> difference that we have just kept doing because "that's how you do it").
>
> Is this code just for us (in which case it is fine) or is it for others
> (in which case we should change '/usr/bin/timeout' to just 'timeout' and
> similar for other command-line commands.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/npad/sidestream/pull/15/files#r37430549>.
>